Tuning library for openai

ajakubo1 opened the pull request:
@jtrain - just a mention for you here so you know what I'm doing with the lib. My intentions right now are:

- write my own output formatted (no monkey-patching needed)
- Adding that ``•`` character to markdown parser (might infolve monkey-patching, but it's not needed - I will be aiming to run a ``replace`` or even writing my own preprocessor
- Changing how ``ListIndentProcessor`` is working - and for that - I think I will need some monkey-patching. Why do I even do it? well, this is because one annoying thing is this thing that I need to do in the test case: ``test_text.replace('•', '  •')``.  If I don't do it, current markdown processor thinks that the unordered list ``dots`` are continuation of ordered list numbering. This is annoying, I strongly feel that this can be reformatted in a more sensible way. It has the numbering, it can KNOW that this unordered list is a sub-list of an ordered list.

I believe all of this is possible to do.

ajakubo1 committed:
Adding a test for sweetprocess (copied from main lib)

ajakubo1 committed:
making sure that the formatters "understand" • sign used by openai

very naughty, chatgpt. Very naughty.

ajakubo1 commented:
```
class UListProcessor(OListProcessor):
    """ Process unordered list blocks. """

    TAG = 'ul'

    def __init__(self, parser):
        super().__init__(parser)
        # Detect an item (``1. item``). ``group(1)`` contains contents of item.
        self.RE = re.compile(r'^[ ]{0,%d}[*+-•][ ]+(.*)' % (self.tab_length - 1))

```

I just realized why mixing ordered and unordered lists without indentations get so confusing outputs :P...

ajakubo1 committed:
tweaking with ul/ol processors that they see only the sibling of their type

ajakubo1 commented:
Current output:

```
<ol>
<li>
<p>Introduce yourself and explain the purpose of the interview.</p>
</li>
<li>
<p>Outline the general structure of the interview and explain that the participant is free to elaborate on any point they feel is important.</p>
</li>
<li>
<p>Ask open-ended questions about the participant's experience of walking as a child. Questions may include:</p>
</li>
</ol>
<ul>
<li>
<p>What were your favorite walking destinations as a child?</p>
</li>
<li>
<p>How did walking make you feel?</p>
</li>
<li>
<p>What did you think about while walking?</p>
</li>
<li>
<p>What were the most memorable walking experiences you had as a child?</p>
</li>
<li>
<p>What was your relationship with walking as a child?</p>
</li>
<li>
<p>How did you and your family use walking as a mode of transportation?</p>
</li>
</ul>
<ol>
<li>
<p>Ask follow-up questions to further explore the participant's experience.</p>
</li>
<li>
<p>Thank the participant for their time.</p>
</li>
</ol>
```

I believe this is pretty close to what we actually need :>.

tbh - maybe this IS enough.

jtrain commented:
I'll check it out, we can roll it out in background if it improves on what we have. while we focus on the next part, getting this into the hands of as many people as possible.

ajakubo1 commented:
Oh, it's not ready for any deploy, I just went through points 2 and 3. I need to write my own output formatter now.

But I believe that what I did is enough for everything to work. We actually don't need those ``ul`` and ``ol`` lists to be consistently at the same level. We just need to check if the element we are trying to render is in ``level 0`` of the document (it's parent is an ``ul`` or a ``ol`` or a ``p`` attached directly to the root element) - if that's the case - it's a single step. If it's not - it's probably a sub-step. So overall - this is not perfect - but it will give us what we actually need.

ajakubo1 commented:
Btw - I do agree that wht I should do is - I should define my own markdown class in sweetprocess. This fork is only for experimenting. I'm pretty sure I will be able to "extract" those two parsers and replace them with my own parsers with different stats for it to work.

ajakubo1 committed:
split the IF (i am going to need it later)

ajakubo1 committed:
took me a while to realize how to sub-list an unindented list

ajakubo1 committed:
adding additional test just for my sanity

ajakubo1 committed:
haha, ok - so this is a bit funny

It looks like - the changes that I have made are not impacting the tests
at all. Like - the most recent changes with the lists (which is - fair).

However - by adding an additional 'parsing' argument - the regexes
started breaking.

This means to me - that I probably need to replace the "dot" char into
a proper - char for lists.

ajakubo1 commented:
An update on my end - I realized that I didn't do a good job yesterday after all. instead of that output https://github.com/SweetProcess/markdown/pull/1#issuecomment-1428194599, I was aiming for this output:

```
<ol>
<li>
<p>Introduce yourself and explain the purpose of the interview.</p>
</li>
<li>
<p>Outline the general structure of the interview and explain that the participant is free to elaborate on any point they feel is important.</p>
</li>
<li>
<p>Ask open-ended questions about the participant's experience of walking as a child. Questions may include:</p>
</li>
<ul>
<li>
<p>What were your favorite walking destinations as a child?</p>
</li>
<li>
<p>How did walking make you feel?</p>
</li>
<li>
<p>What did you think about while walking?</p>
</li>
<li>
<p>What were the most memorable walking experiences you had as a child?</p>
</li>
<li>
<p>What was your relationship with walking as a child?</p>
</li>
<li>
<p>How did you and your family use walking as a mode of transportation?</p>
</li>
</ul>
<li>
<p>Ask follow-up questions to further explore the participant's experience.</p>
</li>
<li>
<p>Thank the participant for their time.</p>
</li>
</ol>
```

(only one ol tag)

this is what I'm doing currently, although - the code is a bit more messy.

ajakubo1 committed:
instead of changing the regexes (scary) - do a replace for dot-notation lists

this should work I think.

ajakubo1 committed:
one more funky test that was problematic earlier

ajakubo1 commented:
OK - once again - the only thing left to do is to write output formatter (this might be a bit annoying based on what I've found already, I really hate working with HTML).

ajakubo1 commented:
after I write the output formatter, I will try to move that whole thing to sweetprocess and run the tests on it and see how it goes.

ajakubo1 committed:
first step in custom output formatter

ajakubo1 committed:
Adding a step converter

ajakubo1 committed:
I hate parsing xml

so the reason for it is - what we want to do, is we want to put
<ul> into <li> of a <ol>. Not <ul> directly as a child of <ol>.
that doesn't make sense.

ajakubo1 committed:
Reformat the text in such a way, that the first sentence is the title, put rest of the stuff into the content.

ajakubo1 committed:
running flat code made me realize that I was ignoring some of the elements which len is 0

ajakubo1 committed:
Tuning the tests so we can see the output right-off the bat

ajakubo1 commented:
@jtrain - the contents I managed to generate from the tests:

```python
[{'content': '',
  'fields': [],
  'files': [],
  'id': 'fdf0cc96fea643839a52dd4cb1495a15',
  'images': [],
  'n': 1,
  'title': 'Introduce yourself and explain the purpose of the interview.',
  'type': 'content'},
 {'content': '',
  'fields': [],
  'files': [],
  'id': 'd2d02cf542154c019b34fb31f3ccc3f7',
  'images': [],
  'n': 2,
  'title': 'Outline the general structure of the interview and explain that '
           'the participant is free to elaborate on any point they feel is '
           'important.',
  'type': 'content'},
 {'content': 'Questions may include: <ul><li><p>What were your favorite '
             'walking destinations as a child?</p></li><li><p>How did walking '
             'make you feel?</p></li><li><p>What did you think about while '
             'walking?</p></li><li><p>What were the most memorable walking '
             'experiences you had as a child?</p></li><li><p>What was your '
             'relationship with walking as a child?</p></li><li><p>How did you '
             'and your family use walking as a mode of '
             'transportation?</p></li></ul>',
  'fields': [],
  'files': [],
  'id': 'fa84bb1613fc48398a4c82ff293533ad',
  'images': [],
  'n': 3,
  'title': "Ask open-ended questions about the participant's experience of "
           'walking as a child.',
  'type': 'content'},
 {'content': '',
  'fields': [],
  'files': [],
  'id': '4017a32d17a54f4aa3dc9e7132f9a204',
  'images': [],
  'n': 4,
  'title': "Ask follow-up questions to further explore the participant's "
           'experience.',
  'type': 'content'},
 {'content': '',
  'fields': [],
  'files': [],
  'id': '6c6a0cff3c204480a25d694a2f0755f5',
  'images': [],
  'n': 5,
  'title': 'Thank the participant for their time.',
  'type': 'content'}]



[{'content': '',
  'fields': [],
  'files': [],
  'id': '6388b26022dd492a8b4092aa53822f1d',
  'images': [],
  'n': 1,
  'title': 'This & that.',
  'type': 'content'},
 {'content': '',
  'fields': [],
  'files': [],
  'id': '1479a7ea3f5148d6877ace35ab13e687',
  'images': [],
  'n': 2,
  'title': '4 < 5.',
  'type': 'content'},
 {'content': '',
  'fields': [],
  'files': [],
  'id': 'ae84d37067144450a4da67f4582e491b',
  'images': [],
  'n': 3,
  'title': '6 > 5.',
  'type': 'content'}]



[{'content': '',
  'fields': [],
  'files': [],
  'id': 'c73948761f1f4e9da1ebdfb0f776e8f2',
  'images': [],
  'n': 1,
  'title': 'You might want to start with this',
  'type': 'content'},
 {'content': '',
  'fields': [],
  'files': [],
  'id': '6c649496246f4225a778d7988d9b48ea',
  'images': [],
  'n': 2,
  'title': 'one',
  'type': 'content'},
 {'content': '',
  'fields': [],
  'files': [],
  'id': '2204a3c0bd194fd9a99bd45690a657f5',
  'images': [],
  'n': 3,
  'title': 'two',
  'type': 'content'},
 {'content': 'I have no idea what is going on here.',
  'fields': [],
  'files': [],
  'id': '63f978c459ea46359e255714d0413c8f',
  'images': [],
  'n': 3,
  'title': 'What is going on in here?',
  'type': 'content'},
 {'content': '',
  'fields': [],
  'files': [],
  'id': '28ffafe292904fe78ae02f7b96ecf0e3',
  'images': [],
  'n': 4,
  'title': 'two',
  'type': 'content'},
 {'content': '',
  'fields': [],
  'files': [],
  'id': 'e5571f91c2b6403c910876f184b20433',
  'images': [],
  'n': 5,
  'title': 'three',
  'type': 'content'},
 {'content': '',
  'fields': [],
  'files': [],
  'id': 'd67fdae7ccdb42f79006d359c9fba70d',
  'images': [],
  'n': 5,
  'title': 'Then you migth want to do this',
  'type': 'content'},
 {'content': '',
  'fields': [],
  'files': [],
  'id': 'b4740738f4d245b38fb5827ae437bee5',
  'images': [],
  'n': 6,
  'title': 'this',
  'type': 'content'},
 {'content': '',
  'fields': [],
  'files': [],
  'id': '2d9f8cbf53a247459c2eef23a63a9ddb',
  'images': [],
  'n': 7,
  'title': 'that',
  'type': 'content'}]

```

I believe this is quite decent :>...

ajakubo1 commented:
Just checking in the raw content if the chosen format makes sense:

![image](https://user-images.githubusercontent.com/8577405/218791909-6b1325f2-4b4a-4146-81ea-51dfd740049e.png)

@jtrain - I believe I will create that markdown formatter in sweetprocess branch tomorrow. I think it will work nicely.
